### PR TITLE
[SPARK-27707][SQL] Prune unnecessary nested fields from Generate

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Performance issue using explode was found when a complex field contains huge array is to get duplicated as the number of exploded array elements. Given example:

```scala
val df = spark.sparkContext.parallelize(Seq(("1",
  Array.fill(M)({
    val i = math.random
    (i.toString, (i + 1).toString, (i + 2).toString, (i + 3).toString)
  })))).toDF("col", "arr")
  .selectExpr("col", "struct(col, arr) as st")
  .selectExpr("col", "st.col as col1", "explode(st.arr) as arr_col")
```

The explode causes `st` to be duplicated as many as the exploded elements.

Benchmarks it:

```
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.14.4
[info] Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
[info] generate big nested struct array:         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] generate big nested struct array wholestage off          52668          53162         699          0.0      877803.4       1.0X
[info] generate big nested struct array wholestage on          47261          49093        1125          0.0      787690.2       1.1X
[info]
```

The query plan:
```
== Physical Plan ==
 Project [col#508, st#512.col AS col1#515, arr_col#519]
 +- Generate explode(st#512.arr), [col#508, st#512], false, [arr_col#519]
    +- Project [_1#503 AS col#508, named_struct(col, _1#503, arr, _2#504) AS st#512]
       +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._1, true, false) AS _1#503, mapobjects(MapObjects_loopValue84, MapObjects_loopIsNull84,      ObjectType(class scala.Tuple4), if (isnull(lambdavariable(MapObjects_loopValue84, MapObjects_loopIsNull84, ObjectType(class scala.Tuple4), true)))     null else named_struct(_1, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue84, MapObjects_loopIsNull84, ObjectType(class scala.Tuple4), true))._1, true, false), _2, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue84, MapObjects_loopIsNull84, ObjectType(class scala.Tuple4), true))._2, true, false), _3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String,     StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue84, MapObjects_loopIsNull84, ObjectType(class scala.Tuple4), true))._3, true,  false), _4, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue84,   MapObjects_loopIsNull84, ObjectType(class scala.Tuple4), true))._4, true, false)), knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._2, None) AS _2#504]
          +- Scan[obj#534]
```

This patch takes nested column pruning approach to prune unnecessary nested fields. It adds a projection of the needed nested fields as aliases on the child of `Generate`, and substitutes them by alias attributes on the projection on top of `Generate`.

Benchmarks it after the change:
```
 [info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.14.4
 [info] Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
 [info] generate big nested struct array:         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 [info] ------------------------------------------------------------------------------------------------------------------------
 [info] generate big nested struct array wholestage off            311            331          28          0.2        5188.6       1.0X
 [info] generate big nested struct array wholestage on            297            312          15          0.2        4947.3       1.0X
 [info]
```

The query plan:
```
== Physical Plan ==
 Project [col#592, _gen_alias_608#608 AS col1#599, arr_col#603]
 +- Generate explode(st#596.arr), [col#592, _gen_alias_608#608], false, [arr_col#603]
    +- Project [_1#587 AS col#592, named_struct(col, _1#587, arr, _2#588) AS st#596, _1#587 AS _gen_alias_608#608]
       +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(in
 put[0, scala.Tuple2, true]))._1, true, false) AS _1#587, mapobjects(MapObjects_loopValue102, MapObjects_loopIsNull102, ObjectType(class scala.Tuple4),
 if (isnull(lambdavariable(MapObjects_loopValue102, MapObjects_loopIsNull102, ObjectType(class scala.Tuple4), true))) null else named_struct(_1,        staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue102,              MapObjects_loopIsNull102, ObjectType(class scala.Tuple4), true))._1, true, false), _2, staticinvoke(class org.apache.spark.unsafe.types.UTF8String,    StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue102, MapObjects_loopIsNull102, ObjectType(class scala.Tuple4), true))._2,      true, false), _3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString,                                                 knownnotnull(lambdavariable(MapObjects_loopValue102, MapObjects_loopIsNull102, ObjectType(class scala.Tuple4), true))._3, true, false), _4,            staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(lambdavariable(MapObjects_loopValue102,              MapObjects_loopIsNull102, ObjectType(class scala.Tuple4), true))._4, true, false)), knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._2,      None) AS _2#588]
          +- Scan[obj#586]
```

This behavior is controlled by a SQL config `spark.sql.optimizer.expression.nestedPruning.enabled`.

## How was this patch tested?

Added benchmark.

### CODE REVIEW

**Comment by viirya** [comment]
cc @uzadude @cloud-fan @dongjoon-hyun

**Comment by uzadude** [comment]
@viirya - looks great! exactly what I had in mind but wasn't sure how to implement it.

**Comment by viirya** [comment]
retest this please.

**Comment by dongjoon-hyun** [comment]
cc @dbtsai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 1]
```diff
@@ -32,9 +32,7 @@ object NestedColumnAliasing {
 
   def unapply(plan: LogicalPlan)
     : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = plan match {
-    case Project(projectList, child)
-        if SQLConf.get.nestedSchemaPruningEnabled && canProjectPushThrough(child) =>
-      getAliasSubMap(projectList)
+    case Project(projectList, child) => getAliasSubMap(projectList)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: @viirya . Sorry, but this is a regression on all the existing code. We should avoid `getAliasSubMap` invocation. https://github.com/apache/spark/pull/24637/files#diff-a636a87d8843eeccca90140be91d4fafR635 doesn't prevent `getAliasSubMap` invocation inside `unapply`.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: I see. If so, I need to make a little change to prevent it. Will change it later.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala [hunk 2]
```diff
@@ -130,6 +130,18 @@ object MiscBenchmark extends SqlBasedBenchmark {
...
+        import spark.implicits._
+        val df = spark.sparkContext.parallelize(Seq(("1",
+          Array.fill(M)({
+            val i = math.random
+            (i.toString, (i + 1).toString, (i + 2).toString, (i + 3).toString)
+          })))).toDF("col", "arr")
+          .selectExpr("col", "struct(col, arr) as st")
+          .selectExpr("col", "st.col as col1", "explode(st.arr) as arr_col")
+        df.collect()
+      }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala]: Please add benchmark result in this PR.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala]: This one, `sql/core/benchmarks/MiscBenchmark-results.txt`~

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala]: Yea, I posted the benchmark result for the added one. Because I think the all benchmark results should be run on r3.xlarge machine, I plan to post it later.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/MiscBenchmark.scala]: Will update `MiscBenchmark-results` soon.

File: sql/core/benchmarks/MiscBenchmark-results.txt [hunk 3]
```diff
@@ -2,119 +2,126 @@
...
+OpenJDK 64-Bit Server VM 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03 on Linux 4.15.0-1021-aws
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
-generate explode map:                    Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------
-generate explode map wholestage off         22128 / 22578          0.8        1318.9       1.0X
-generate explode map wholestage on          16421 / 16520          1.0         978.8       1.3X
+generate explode map:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+generate explode map wholestage off               50364          50656         413          0.3        3001.9       1.0X
+generate explode map wholestage on                43484          43833         313          0.4        2591.9       1.2X
```
Comment (viirya) [inline sql/core/benchmarks/MiscBenchmark-results.txt]: This looks regression from last committed benchmark results. I don't think this PR can cause it. So to verify it, I ran this benchmark on current master branch:

```
[info] OpenJDK 64-Bit Server VM 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03 on Linux 4.15.0-1021-aws
[info] Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
[info] generate explode array:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] generate explode array wholestage off             15893          15991         138          1.1         947.3       1.0X
[info] generate explode array wholestage on              10798          10833          36          1.6         643.6       1.5X
[info] 
[info] Running benchmark: generate explode map
[info]   Running case: generate explode map wholestage off
[info]   Stopped after 2 iterations, 102915 ms
[info]   Running case: generate explode map wholestage on
[info]   Stopped after 5 iterations, 216755 ms
[info] 
[info] OpenJDK 64-Bit Server VM 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03 on Linux 4.15.0-1021-aws
[info] Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
[info] generate explode map:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] generate explode map wholestage off               50273          51458        1676          0.3        2996.5       1.0X
[info] generate explode map wholestage on                43113          43351         278          0.4        2569.8       1.2X
[info] 
```

The all benchmark result can be found at https://gist.github.com/viirya/c850b49e364e8e6aaa6271ae9e3343fc

Comment (viirya) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: I open https://issues.apache.org/jira/browse/SPARK-27779 to track it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -588,6 +588,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
         .map(_._2)
       p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices))
 
+    // prune unrequired nested fields
+    case p @ Project(projectList, g: Generate) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Why we need to special case `Generate`? I see there is a general case for nested column pruning below.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think this is more general case you are looking for:

https://github.com/apache/spark/blob/163a6e298213f216f74f4764e241ee6298ea30b6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L623-L624

In the general case, it's doing pruning only when the flag (nestedSchemaPruningEnabled) is enabled. The case considers some operators that nested project can be pushed through. `Generate` isn't one of them. So the general case doesn't work on this case.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: `Generate` is special for another reason. We can't prune an output from its child even just a nested field of the output is used in the top project list. The generator could use it.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Shall we add `if SQLConf.get.nestedSchemaPruningEnabled`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala [hunk 5]
```diff
@@ -264,9 +266,10 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
       .analyze
     comparePlans(optimized, expected)
   }
+}
 
-
-  private def collectGeneratedAliases(query: LogicalPlan): ArrayBuffer[String] = {
+object NestedColumnAliasingSuite {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala]: Yep. This looks like the minimal challenge for now.

File: sql/core/benchmarks/MiscBenchmark-results.txt [hunk 6]
```diff
@@ -2,119 +2,126 @@
...
+OpenJDK 64-Bit Server VM 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03 on Linux 4.15.0-1021-aws
 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
-range/limit/sum:                         Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------
-range/limit/sum wholestage off                 229 /  236       2288.9           0.4       1.0X
-range/limit/sum wholestage on                  257 /  267       2041.0           0.5       0.9X
+range/limit/sum:                          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+range/limit/sum wholestage off                      191            205          19       2738.4           0.4       1.0X
+range/limit/sum wholestage on                       112            124          13       4699.4           0.2       1.7X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/MiscBenchmark-results.txt]: Ur, this is irrelevant, but the ratio looks weird.

File: sql/core/benchmarks/MiscBenchmark-results.txt [hunk 7]
```diff
@@ -2,119 +2,126 @@
...
-generate big struct array wholestage off       708 /  776          0.1       11803.5       1.0X
-generate big struct array wholestage on        535 /  589          0.1        8913.9       1.3X
+generate big struct array:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+generate big struct array wholestage off            663            705          59          0.1       11054.4       1.0X
+generate big struct array wholestage on             576            612          34          0.1        9602.9       1.2X
+
+OpenJDK 64-Bit Server VM 1.8.0_212-8u212-b03-0ubuntu1.18.04.1-b03 on Linux 4.15.0-1021-aws
+Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
+generate big nested struct array:         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/MiscBenchmark-results.txt]: Shall we add this result only?

Comment (viirya) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: Manually edit `MiscBenchmark-results.txt `?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: `nestedSchemaPruningEnabled` is for pruning nested fields from logical relation. But this fix isn't due to the same cause. For the data sources can't be pruned nested fields, it is also useful to apply this fix.

File: sql/core/benchmarks/MiscBenchmark-results.txt (hunk 7 above)
Comment (viirya) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: @dongjoon-hyun ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Out of curiosity, why don't we need another configuration then if `spark.sql.optimizer.nestedSchemaPruning.enabled` isn't part of this optimization, or fixing the doc of `spark.sql.optimizer.nestedSchemaPruning.enabled`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This looks a general fix no matter the setting of `nestedSchemaPruning` is. Do we need a config to disable this?

File: sql/core/benchmarks/MiscBenchmark-results.txt (hunk 7 above)
Comment (viirya) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: @dongjoon-hyun Any advice for this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: My impression was that we need a configuration but I think you or @dongjoon-hyun have more context then me about nested pruning stuff. @cloud-fan, @dongjoon-hyun, @gatorsmile, can you make a call here if we need a config or not?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: `nestedSchemaPruning` is vague as it can point to nested pruning in scan, or general nested pruning in other operators. Currently the config affects scan nested pruning. I feel it is not tightly related to the fix here, because the fix isn't for scan.

If we are considering a config here, I think a different config or fix the doc of `nestedSchemaPruning` to explicitly indicate it also helps general nested pruning.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Let me fix the doc of `nestedSchemaPruning` and apply this pruning when `nestedSchemaPruning` is enabled.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 8]
```diff
@@ -588,6 +588,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
         .map(_._2)
       p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices))
 
+    // prune unrequired nested fields
+    case p @ Project(projectList, g: Generate) if SQLConf.get.nestedSchemaPruningEnabled =>
+      NestedColumnAliasing.getAliasSubMap(projectList).map {
+        case (nestedFieldToAlias, attrToAliases) =>
+          val newChild = g.withNewChildren(g.children.map { child =>
+            Project(child.output ++ attrToAliases.values.flatten, child)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This looks addition instead of a replacing. So, this may optimize `explode`, but we cannot avoid the full scan of underlying parquet files?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -1576,7 +1576,8 @@ object SQLConf {
       .doc("Prune nested fields from a logical relation's output which are unnecessary in " +
         "satisfying a query. This optimization allows columnar file format readers to avoid " +
         "reading unnecessary nested column data. Currently Parquet and ORC are the " +
-        "data sources that implement this optimization.")
+        "data sources that implement this optimization. This optimization also allows pruning " +
+        "unnecessary nested fields in expressions of operator.")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Can we have another configuration for this `pruning on expressions`? For me, this is not about `nestedSchemaPruning` because this seems irrelevant `schema` changing.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: We talked about it in previous discussion, that is an option. Reusing `nestedSchemaPruning` was to simplify configs. Let me create a separate config for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 8 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This is good point. I was planning to deal this as separate issue. As you mentioned and we control it with a pruning on expressions config now, I think it is good to deal with it here. I will update.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 9 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Sorry about asking this. During last review, I realized that this can disable all the other nested schema pruning features to the users.

File: sql/core/benchmarks/MiscBenchmark-results.txt (hunk 7 above)
Comment (dongjoon-hyun) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: Sorry for being late. Instead, for this one, I'll make a PR to you.

**Comment by dongjoon-hyun** [comment]
Oh, I faced the following during benchmark. It's weird. Let me rerun it.
```
[info] 22:18:13.279 WARN org.apache.spark.network.server.TransportChannelHandler: Exception in connection from /172.31.20.31:54998
[info] java.util.concurrent.RejectedExecutionException: event executor terminated
```

File: sql/core/benchmarks/MiscBenchmark-results.txt [hunk 10]
```diff
@@ -2,119 +2,126 @@
...
+------------------------------------------------------------------------------------------------------------------------
+generate big nested struct array wholestage off            635            683          68          0.1       10582.3       1.0X
+generate big nested struct array wholestage on            647            679          38          0.1       10785.4       1.0X
```
Comment (dongjoon-hyun) [inline sql/core/benchmarks/MiscBenchmark-results.txt]: Ur, I got a wrong result on this. I rebased this PR to the master.
```
-generate big nested struct array wholestage off            635            683          68          0.1       10582.3       1.0X
-generate big nested struct array wholestage on            647            679          38          0.1       10785.4       1.0X
+generate big nested struct array wholestage off          57843          62617        2949          0.0      964042.1       1.0X
+generate big nested struct array wholestage on          49793          53627         NaN          0.0      829888.7       1.2X
```

Comment (dongjoon-hyun) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: I'll make a PR to you for comparison.

**Comment by dongjoon-hyun** [comment]
This is the result on this PR
- https://github.com/dongjoon-hyun/spark/commit/e41f17e034970bd43d6397d73b366d51a246f596

The new configuration seems to be missed on the benchmark. Do you want to enable the configuration for the new benchmark case?

**Comment by viirya** [comment]
@dongjoon-hyun oh, yes, the new config should be enabled. Let me add it. Would you like to re-submit a PR after adding the config?

**Comment by dongjoon-hyun** [comment]
Sure. I'll rerun and make a PR to you tonight.

**Comment by dongjoon-hyun** [comment]
Weirdly. The benchmark suite failed with the exact same reason at the first run. I'm running the second run.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 11]
```diff
@@ -1589,6 +1589,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val NESTED_PRUNING_ON_EXPRESSIONS =
+    buildConf("spark.sql.optimizer.expression.nestedPruning.enabled")
+    .internal()
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: indentation?

File: sql/core/benchmarks/MiscBenchmark-results.txt (hunk 6 above)
Comment (dongjoon-hyun) [reply sql/core/benchmarks/MiscBenchmark-results.txt]: This seems to be improved by some reason. This is consistently better in @viirya and my tests.
```
[info] OpenJDK 64-Bit Server VM 1.8.0_212-b04 on Linux 3.10.0-862.3.2.el7.x86_64
[info] Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
[info] range/limit/sum:                          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] range/limit/sum wholestage off                      222            226           6       2359.5           0.4       1.0X
[info] range/limit/sum wholestage on                       114            121           8       4608.4           0.2       2.0X
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 11 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Oops, thanks for the catch.

**Comment by dongjoon-hyun** [comment]
I made a PR to you which reverts all the other benchmark result to the master and adds the newly added test case result.
- https://github.com/viirya/spark-1/pull/7

**Comment by viirya** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
@viirya . The PR looks fine. The only concern is that this PR will match the other `Generator` (not only `explode`. We have two options.
1. Catching for only `Explode`.
2. Adding more test case for the other `Generator`.

First of all, I remove `in explode` from the title because this PR works for the others too (AS-IS).
Which one do you want to proceed? I'm fine for both one if this PR is clear.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun for the advice! I will add more few test cases targeting other Generator.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @viirya !

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruningSuite.scala [hunk 12]
```diff
@@ -101,6 +103,56 @@ class ColumnPruningSuite extends PlanTest {
...
+      }
+    }
+
+    runTest(
+      Explode('c.getField("e")),
+      aliases => Explode($"${aliases(1)}".as("c.e")),
+      aliases => Seq('c.getField("d").as(aliases(0)), 'c.getField("e").as(aliases(1))),
+      Seq(2)
+    )
+    runTest(Stack(2 :: 'c.getField("f") :: 'c.getField("g") :: Nil),
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruningSuite.scala]: @dongjoon-hyun I only added `Stack` for test.

Comment (dongjoon-hyun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruningSuite.scala]: Thank you for adding more tests, @viirya !

**Comment by dongjoon-hyun** [comment]
Thank you for adding a new test. Are you going to add more tests since `Stack` is one of them? In fact, we need more to be exhaustive.

One another approach is simply reducing the scope to the original goal. We can match only-`Explode` like the following in this PR.
```
case p @ Project(projectList, g: Generate) if ...
case p @ Project(projectList, g @ Generate(_: Explode, _, _, _, _, _)) if ...
```

Later, to cover more patterns, I think we need `unapply` and a white-list approach like the following.
```
  private def canProjectPushThrough(plan: LogicalPlan) = plan match {
    case _: GlobalLimit => true
    ...
    case _ => false
  }
```

Since this PR is here for a long time, how about finishing here with `Explode` first, @viirya ?

**Comment by viirya** [comment]
@dongjoon-hyun I added more generators. I think existing generators should be found in the test.

**Comment by viirya** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for adding more. Then, please do a white list approach for those four expressions as I mentioned here.
- https://github.com/apache/spark/pull/24637#issuecomment-512358172

cc @cloud-fan and @gatorsmile

**Comment by viirya** [comment]
I'm fine to add a white-list, thro I think this approach is not generator-specific. It is more conservative and safer, anyway.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. 
Thank you so much for working on this.
Merged to master.

**Comment by cloud-fan** [comment]
We hit an exception caused by this rule. The plan becomes invalid after optimization
```
+- *(2) !Project [_gen_alias_68718#68718L AS cardinality#68575L, _gen_alias_68719#68719 AS durationSec#68576, _gen_alias_68720#68720 AS group#68578, _gen_alias_68721#68721 AS jobUuid#68579, _gen_alias_68722#68722 AS suite#68584, _gen_alias_68723#68723 AS testcase#68585, sha1(cast(_gen_alias_68721#68721 as binary)) AS jobSha#68600, sha1(cast(concat(_gen_alias_68722#68722, -, _gen_alias_68720#68720, -, cast(_gen_alias_68718#68718L as string), -, _gen_alias_68723#68723) as binary)) AS caseSha#68615]
      +- *(2) Generate explode(results#64594), false, [flattenRuns#68572]
         +- *(2) Project [results#64594]
            +- *(2) Sort [startTime#68717 DESC NULLS LAST], true, 0
```

We generate `_gen_alias` attributes in the parent `Project` but they are not available in the child `Generate`.

@viirya Can you help to take a look? thanks!

**Comment by viirya** [comment]
@cloud-fan Yea, will look at it tomorrow. Do you have test case? If no, I may try to reproduce it.

**Comment by cloud-fan** [comment]
It's a very long query and I'm trying to minimize. Let's see if there is some clue about the query plan.

**Comment by dongjoon-hyun** [comment]
Thank you for reporting, @cloud-fan !

**Comment by dongjoon-hyun** [comment]
Could you file a JIRA for that, @cloud-fan ?

**Comment by viirya** [comment]
Is `flattenRuns` (generatorOutput) also not in the parent Project?

**Comment by viirya** [comment]
Re-checked the current rule. Still cannot find clue from it and and above query plan. At first glance, I suspect if there are nested column access at the top Project but not in Generate. I tested it and re-checked the rule, looks it is good. @cloud-fan May you have more clues?
